strawberryfields/circuitspecs/chip0.py

```diff
@@ -51,7 +62,134 @@ class Chip0Specs(CircuitSpecs):
+
+        # get circuit registers
+        regrefs = {q for cmd in B for q in cmd.reg}
+
+        if len(regrefs) != self.modes:
+            raise CircuitError("S2gates do not appear on the correct modes.")
+
+        # Compile the unitary: combine and then decompose all unitaries
+        # -------------------------------------------------------------
+        A, B, C = group_operations(seq, lambda x: isinstance(x, (ops.Rgate, ops.BSgate)))
+
+        # begin unitary lists for mode [0, 1] and modes [2, 3] with
+        # two identity matrices. This is because multi_dot requires
+        # at least two matrices in the list.
+        U_list01 = [np.identity(self.modes // 2, dtype=np.complex128)] * 2
+        U_list23 = [np.identity(self.modes // 2, dtype=np.complex128)] * 2
+
+        if not B:
+            # no interferometer was applied
+            A, B, C = group_operations(seq, lambda x: isinstance(x, ops.S2gate))
+            A = B  # move the S2gates to A
+        else:
+            for cmd in B:
+                # calculate the unitary matrix representing each
+                # rotation gate and each beamsplitter
+                # Note: this is done separately on modes [0, 1]
+                # and modes [2, 3]
+                modes = [i.ind for i in cmd.reg]
+                params = [i.x for i in cmd.op.p]
+                U = np.identity(self.modes // 2, dtype=np.complex128)
+
+                if isinstance(cmd.op, ops.Rgate):
+                    m = modes[0]
+                    U[m % 2, m % 2] = np.exp(1j * params[0])
+
+                elif isinstance(cmd.op, ops.BSgate):
+                    m, n = modes
+
+                    t = np.cos(params[0])
+                    r = np.exp(1j * params[1]) * np.sin(params[0])
+
+                    U[m % 2, m % 2] = t
+                    U[m % 2, n % 2] = -np.conj(r)
+                    U[n % 2, m % 2] = r
+                    U[n % 2, n % 2] = t
+
+                if set(modes).issubset({0, 1}):
+                    U_list01.insert(0, U)
+                elif set(modes).issubset({2, 3}):
+                    U_list23.insert(0, U)
+                else:
+                    raise CircuitError(
+                        "Unitary must be applied separately to modes [0, 1] and modes [2, 3]."
+                    )
+
+        # multiply all unitaries together, to get the final
+        # unitary representation on modes [0, 1] and [2, 3].
+        U01 = multi_dot(U_list01)
+        U23 = multi_dot(U_list23)
+
+        # check unitaries are equal
+        if not np.allclose(U01, U23):
+            raise CircuitError(
+                "Interferometer on modes [0, 1] must be identical to interferometer on modes [2, 3]."
+            )
+
+        U = block_diag(U01, U23)
+
+        # replace B with an interferometer
+        B = [
+            Command(ops.Interferometer(U01), registers[:2]),
+            Command(ops.Interferometer(U23), registers[2:]),
```

Comment:
Is this the only place where `registers` is needed? Would it be possible to deduce the correct RegRefs by looking at the `Command` instances in `seq`?

Reply:
I ran into an issue where a circuit with no unitary (which should compile fine), I could not apply an identity `Interferometer` decomposition because I did not have access to the `regref` instances. While I could have accessed the regref instances from the `MeasureFock` commands, this felt hacky, and there may be more use cases in the future (e.g., compiling an empty program).